### ISSUE TITLE
backfill from the same format as "promtool tsdb dump"

### ISSUE THREAD

**Issue by ArthurSens** [prometheus/prometheus#8281]
## Proposal
We can also data from Prometheus with `promtool tsdb dump`, but unfortunately we can only import it back in OpenMetrics format so we can't re-use this file to import that data into another Prometheus server. The proposal is to add support for creating Prometheus compatible blocks from the same format that we currently dump, so we can move data from different prometheus easily.

**Use case. Why is this important?**
Where I work, a decision was made in the past that we'd have one Prometheus server for each application/exporter and it's not scaling too well. We'd like to consolidate all that data into a single Prometheus that will now scrape all our metrics endpoints. We want to shut down several Prometheus, but we'd like to keep their data in our new centralized one.

With this proposal, it will be a lot easier to dump and import data from one prometheus instance to another.

**Comment by ArthurSens** [comment]
Very similar to #8280, different approaches for the same goal.

**Comment by brian-brazil** [comment]
Why not use the blocks directly off disk? It'll be far faster and more efficient, as you don't need to convert it to a less efficient format just to convert it back.

**Comment by ArthurSens** [comment]
Yes, it is possible to just copy whole tsdb blocks, or even a snapshot to copy the whole tsdb, but with dump/import I can fine grain and get exactly what I want.

Today we can already filter dumped data by timestamp, and I'm hopping that we can add more filters to the dump subcommand, e.g., by labels.

**Comment by ArthurSens** [comment]
Created #8320 for this one, still needs some work to add a subcommand to promtool that uses that parser.

Well, if the idea is really going to be accepted of course :stuck_out_tongue:

**Comment by beorn7** [comment]
Yeah, dumping just certain time slices (as we can do it already) and a subset of time series (a possible new feature of the dump command) and then backfilling it into a different Prometheus instance would be nice. (Currently, users would probably accomplish the latter feature by `grep`ing the result of the current dump command.)

As discussed in #8320, we could have an option to dump to blocks (in a separate directory) directly.

**Comment by brian-brazil** [comment]
One of Thanos/Cortex is working on a tsdb block rewriter, so that plus a snapshot is likely the best way to deal with that efficiently. Once that is complete, we could look at moving it into promtool.

**Comment by beorn7** [comment]
Sounds like a long way to go…

I expect people generally not to be satisfied with OM as the only backfilling option (because tools producing OM file format aren't exactly widespread at the moment, and also for the efficiency reasons discussed in #8320). They will ask for CSV as the "obvious" choice. Once we have explained them that the CSV data model is really at odds with the Prometheus data model, I expect JSON as the next demand.

Perhaps defining a relatively stable JSON representation closely matching the way samples are saved in TSDB (including stale markers) is not such a bad idea. It could be produced and read ad-hoc via JSON tooling by non-Prometheus-related software, and we could support dumping and backfilling it (which could even help overcome versioning problems, like backporting data from a newer to an older Prometheus instance).

While it is nice in principle to have a standard like OM, I get the feeling that we shouldn't shoehorn it into a bulk metrics-data exchange format.

**Comment by brian-brazil** [comment]
It's not that far away for the main bits I gather, full relabelling is trickier is you can't exactly do that in one-pass with constant memory.

I agree that shoehorning OM isn't the right thing here, even ignoring that it can't fully represent this data.

This is all getting off topic though, as this issue is not about general backfilling. This issue is based on the idea that some users might want to dump some data from a Prometheus to then load it into another Prometheus. For that I think blocks are a much better choice than any text-ish format, and we should not prematurely co-mingle that use case, fully general backfilling, and dumping for debugging.

**Comment by ArthurSens** [comment]
> This is all getting off topic though, as this issue is not about general backfilling. This issue is based on the idea that some users might want to dump some data from a Prometheus to then load it into another Prometheus. For that I think blocks are a much better choice than any text-ish format, and we should not prematurely co-mingle that use case, fully general backfilling, and dumping for debugging.

Yep, totally agree that just dumping blocks and move them around is the best approach for dump-undump. The logic sounds kinda funny but fairly efficient, it completely kills the need for the new `create-blocks-from` subcommand

But as beorn said on #8320, maybe we could think of a way to dump something that is not a black-box such as a TSDB block. The dev-summit consensus _"For data science applications, we encourage the creation of tooling to export data from TSDB to a format that is readable by Python & Java."_ may be explored a little bit with the dump subcommand.

Anyway, I'll try to finally start a design doc around the idea and post it here once I'm happy with it. I'll make clear that one approach might not exclude the other (dumping blocks or a formatted file) since they can be used for different cases.

**Comment by brian-brazil** [comment]
For the sake of clarity, such tooling would likely be outside of this repository. The existing remote read endpoint should provide what's needed already to integrate such tooling.

**Comment by ArthurSens** [comment]
https://docs.google.com/document/d/13_wMfTKEWsbIbMChOinKcYR6PW1AyPeCxG-XCLbiV7w/edit#

Alright, I think there is enough info in this document for people to understand all possible approaches and their advantages/disadvantages. For the sake of including everything, I've also added approaches that were already discarded at some point in the past. 

I've also added some implementation ideas, but I'm not really the most experienced person with Prometheus codebase, so if I'm leading in the wrong direction, please let me know :)

**Comment by accerqueira** [comment]
I tried to (naively) filter/convert the dump to OpenMetrics format, like this:
```shell
cat <<EOF | sh | promtool tsdb create-blocks-from openmetrics /dev/stdin
promtool tsdb dump /prometheus/ \
    | grep '<label>="<whatever>"' \
    | sed -r -e 's#\{__name__="([^"]+)", #\1{#' -e 's#, #,#g'
echo -n '# EOF'
EOF
```
but got: `mmap, size 0: invalid argument`.
I guess mmap is used here for performance reasons (is there?), but makes streaming input impossible.

I could break that in 2 steps, but the intermediate OpenMetric file would be huge...

**Comment by beorn7** [comment]
Hello from the bug scrub.

To clarify: The linked design doc proposes to implement a dump of whole blocks rather than implementing an "un-dump" of the `promtool tsdb dump` output. See the docs for all the details.

Changing the title accordingly.

**Comment by beorn7** [comment]
Also note that there is by now an experimental `dump-openmetrics` command, which can be used for in imperfect roundtrip. OpenMetrics doesn't support native histograms yet (but at least that is planned for the future) and it doesn't support staleness markers (which OM will most likely never support because it doesn't have a place in an exposition format).

**Comment by sun105253** [comment]
> 是的，可以只复制整个 tsdb 块，甚至可以复制快照来复制整个 tsdb，但是使用 dump/import 我可以细粒度地得到我想要的东西。
> 
> 今天我们已经可以通过时间戳过滤转储的数据，我希望我们可以向 dump 子命令添加更多过滤器，例如通过指标名称。

  

> Yes, it is possible to just copy whole tsdb blocks, or even a snapshot to copy the whole tsdb, but with dump/import I can fine grain and get exactly what I want.
> 
> Today we can already filter dumped data by timestamp, and I'm hopping that we can add more filters to the dump subcommand, e.g., by metrics name.

I'm currently working on the scenario you mentioned. I have two points that are not very clear to me. How can I integrate the historical data from several Prometheus instances into one? And how can I ensure that this single Prometheus retains its current data during the data migration?
